File: GameOfLife-XCode/SKGameOfLife/GameOfLife.swift
//
//  GameOfLife.swift
//  ShakespeareLogic
//
//  Created by Max Bilbow on 02/07/2015.
//  Copyright © 2015 Rattle Media. All rights reserved.
//

import Foundation

// Field represents a two-dimensional field of cells.
struct Field {
    var s: [[Bool]]
    var w, h: Int
    
    // NewField returns an empty field of the specified width and height.
    static func new(width w: Int, height h: Int) -> Field {
        var result: [[Bool]] = [[Bool]]()
        for var i = 0; i < h; ++i {
            var boolArray = [Bool]()
            for var j = 0; j < w; ++j {
                boolArray.append(false)
            }
            result.append(boolArray)
        }
        return Field(s: result, w: w, h: h)
    }
    
    // Set sets the state of the specified cell to the given value.
    mutating func set(x: Int, y: Int, b: Bool) {
        self.s[y][x] = b
    }
    
    // Alive reports whether the specified cell is alive.
    // If the x or y coordinates are outside the field boundaries they are wrapped
    // toroidally. For instance, an x value of -1 is treated as width-1.
    func alive(var x: Int, var y: Int) -> Bool {
        x += self.w
        x %= self.w
        y += self.h
        y %= self.h
        return self.s[y][x]
    }
    
    // Next returns the state of the specified cell at the next time step.
    func next(x: Int, y: Int) -> Bool {
        // Count the adjacent cells that are alive.
        var alive = 0
        for var i = -1; i <= 1; i++ {
            for var j = -1; j <= 1; j++ {
                if (j != 0 || i != 0) && self.alive(x+i, y: y+j) {
                    alive++
                }
            }
        }
        // Return next state according to the game rules:
        //   exactly 3 neighbors: on,
        //   exactly 2 neighbors: maintain current state,
        //   otherwise: off.
        return alive == 3 || alive == 2 && self.alive(x, y: y)
    }
    
}

// Life stores the state of a round of Conway's Game of Life.
class Life : CustomStringConvertible {
    var a, b: Field
    var w, h: Int
    
    init(fieldA a: Field, fildB b: Field, width w: Int, height h: Int){
        self.a = a
        self.b = b
        self.w = w
        self.h = h
    }
    
    
    // NewLife returns a new Life game state with a random initial state.
    class func new(width: CGFloat, height: CGFloat) -> Life {
        let w = Int(width); let h = Int(height)
        var a: Field = Field.new(width: w, height: h)
        for var i = 0; i < (w * h / 4); i++ {
            a.set(random() % w,y: random() % h, b: true)
        }
        return Life(fieldA: a, fildB: Field.new(width: w, height: h), width: w, height: h)
    }
    
    // NewLife returns a new Life game state with a random initial state.
    class func new(width w: Int, height h: Int) -> Life {
        var a: Field = Field.new(width: w, height: h)
        for var i = 0; i < (w * h / 4); i++ {
            a.set(random() % w,y: random() % h, b: true)
        }
        return Life(fieldA: a, fildB: Field.new(width: w, height: h), width: w, height: h)
    }
    
    // Step advances the game by one instant, recomputing and updating all cells.
    func step() {
        // Update the state of the next field (b) from the current field (a).
        for var y = 0; y < self.h; y++ {
            for var x = 0; x < self.w; x++ {
                self.b.set(x, y: y, b: self.a.next(x, y: y))
            }
        }
        // Swap fields a and b.
        let temp = self.a
        self.a = self.b
        self.b = temp
    }
    
    // String returns the game board as a string.
    var description: String {
        var buf: String = ""//bytes.Buffer
        for var y = 0; y < self.h; y++ {
            for var x = 0; x < self.w; x++ {
                var b: String = " " //byte(" ")
                if self.a.alive(x, y: y) {
                    b = "*"
                }
                buf += b //.WriteByte(b)
            }
            buf += "\n" //.WriteByte("\n")
        }
        return buf//.String()
    }
}

File: GameOfLife-Unity/Assets/GameOfLife/Scripts/GameOfLife.cs
using UnityEngine;
using System.Collections;

/// Life stores the state of a round of Conway's Game of Life.
/// Adapted from example on Go programming website.
public class GameOfLife  {
	Field a, b;
	int w, h;

	public int width {
		get {
			return this.w;
		}
	}

	public int height {
		get {
			return this.h;
		}
	}

	public Field field {
		get {
			return this.a;
		}
	}
			
	public GameOfLife(Field a, Field b, int w, int h) {
		this.a = a;
		this.b = b;
		this.w = w;
		this.h = h;
	}
	
	/// NewLife returns a new Life game state with a random initial state.
	public static GameOfLife New(int w, int h)  {
		Field a = Field.New(w, h);
		for (int i = 0; i < (w * h / 4); i++){
			int X = Random.Range(0, w);
			int Y = Random.Range(0, h);
			a.Set(X, Y, true);
		}
		return new GameOfLife (a, Field.New (w, h), w, h);
	}
	
	/// Step advances the game by one instant, recomputing and updating all cells.
	public void Step() {
		// Update the state of the next field (b) from the current field (a).
		for (int y = 0; y < this.h; y++) {
			for (int x = 0; x < this.w; x++) {
				this.b.Set(x, y, this.a.Next(x, y));
			}
		}
		// Swap fields a and b.
		var temp = this.a;
		this.a = this.b;
		this.b = temp;
	}
	
	/// String returns the game board as a string.
	public override string ToString() {
		string buf = "";//bytes.Buffer
		for (int y = 0; y < this.h; y++ ){
			for (int x = 0; x < this.w; x++ ){
				string b = " "; //byte(" ")
				if (this.a.Alive(x, y)){
					b = "*";
				}
				buf += b; //.WriteByte(b)
			}
			buf += "\n"; //.WriteByte("\n")
		}
		return buf;//.String()
	}

	public class Field {
		/// Field represents a two-dimensional field of cells.
		bool[,] s;
		int w, h;
		
		public Field(bool[,] s, int w, int h) {
			this.s = s;
			this.w = w;
			this.h = h;
		}
		/// NewField returns an empty field of the specified width and height.
		public static Field New(int w, int h) {
			bool[,] result = new bool[h,w];
			for (int i = 0; i < h; ++i ){
				for (int j = 0; j < w; ++j) {
					result[i,j] = false;
				}
			}
			return new Field(result, w, h);
		}
		
		/// Set sets the state of the specified cell to the given value.
		public void Set(int x, int y, bool b) {
			this.s [y, x] = b;
		}
		
		/// Alive reports whether the specified cell is alive.
		/// If the x or y coordinates are outside the field boundaries they are wrapped
		/// toroidally. For instance, an x value of -1 is treated as width-1.
		public bool Alive(int x, int y) {
			x += this.w;
			x %= this.w;
			y += this.h;
			y %= this.h;
			return this.s[y, x];
		}
		
		/// Next returns the state of the specified cell at the next time step.
		public bool Next(int x, int y) {
			// Count the adjacent cells that are alive.
			int alive = 0;
			for (int i = -1; i <= 1; i++) {
				for (int j = -1; j <= 1; j++) {
					if ((j != 0 || i != 0) && this.Alive(x+i, y+j)) {
						alive++;
					}
				}
			}
			// Return next state according to the game rules:
			//   exactly 3 neighbors: on,
			//   exactly 2 neighbors: maintain current state,
			//   otherwise: off.
			return alive == 3 || alive == 2 && this.Alive(x, y);
		}
		
	}
}